metals/src/main/scala/scala/meta/internal/metals/BuildTargetClasses.scala

```diff
@@ -0,0 +1,79 @@
+package scala.meta.internal.metals
+
+import java.util.concurrent.atomic.AtomicReference
+import ch.epfl.scala.bsp4j.ScalaMainClass
+import ch.epfl.scala.{bsp4j => b}
+import scala.collection.concurrent.TrieMap
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.meta.internal.metals.MetalsEnrichments._
+
+final class BuildTargetClasses(
+    buildServer: () => Option[BuildServerConnection]
+)(implicit val ec: ExecutionContext) {
+  private val index = new TrieMap[b.BuildTargetIdentifier, Classes]()
+
+  def isInitialized(target: b.BuildTargetIdentifier): Boolean =
+    classesOf(target).isInitialized.get()
+
+  def main(target: b.BuildTargetIdentifier): TrieMap[String, ScalaMainClass] =
+    classesOf(target).main
+
+  val onCompiled: BatchedFunction[b.BuildTargetIdentifier, Unit] =
+    BatchedFunction.fromFuture(fetchMainClassesFor)
+
+  private def fetchMainClassesFor(
+      targets: Seq[b.BuildTargetIdentifier]
+  ): Future[Unit] = {
+    for (target <- targets) {
+      classesOf(target).clear()
+    }
+
+    buildServer() match {
+      case Some(connection) =>
+        val parameters = new b.ScalaMainClassesParams(targets.asJava)
+        val task = for {
+          mainClasses <- connection.mainClasses(parameters).asScala
+          _ = initializeMainClasses(mainClasses)
+        } yield
+          for (target <- targets) {
```

Comment:
nit: Please use `targets.foreach { target => ... }` instead. This codebase 

- uses `.foreach` for basic iterations
- uses for comprehensions for non-trivial `map/filter/flatMap` chains
- never uses the `for (a <- b)` because it's not very flexible